### PR TITLE
Localization

### PR DESCRIPTION
**Pull request by vanrossumict**
Hi tabuckner!
Thanks for sharing your work.
A nice step in replacing MomentJs with the much lighter DayJS!
I took the liberty to implement localization to this Angular Material DateAdapter.
Hope you are able to review/test it and of course that you like it and merge it :)

### CODE REVIEW

**Comment by tabuckner** [comment]
@vanrossumict Thank you so much for submitting this! Taking a look now. 😄

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts [hunk 1]
```diff
@@ -135,16 +143,18 @@ export class DayjsDateAdapter extends DateAdapter<Dayjs> {
...
+    return value ? this.dayJs(value).locale(this.locale) : null;
   }
 
   format(date: Dayjs, displayFormat: string): string {
     if (!this.isValid(date)) {
-      throw Error('MomentDateAdapter: Cannot format invalid date.');
+      throw Error('DayjsDateAdapter: Cannot format invalid date.');
     }
-    return date.format(displayFormat);
+    return date.locale(this.locale).format(displayFormat);
```
Comment (tabuckner) [inline projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts]: I wonder if this is required here. Maybe this works better if we keep optional locales in mind:
```suggestion
    return this.locale ? date.locale(this.locale).format(displayFormat) : date.format(displayFormat);
```

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts [hunk 2]
```diff
@@ -203,13 +213,12 @@ export class DayjsDateAdapter extends DateAdapter<Dayjs> {
...
-  // TODO: How to create an invalid Dayjs object?
   invalid(): Dayjs {
-    return;
+    return this.dayJs(null);
   }
 
-  private dayJs(input?: any, format?: string): Dayjs {
-    return this.shouldUseUtc ? dayjs.utc(input, format) : dayjs(input, format);
+  private dayJs(input?: any, format?: string, locale?: string): Dayjs {
+    return this.shouldUseUtc ? dayjs(input, { format: format, locale: locale, utc: this.shouldUseUtc }, locale).utc() : dayjs(input, { format: format, locale: locale }, locale);
```
Comment (tabuckner) [inline projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts]: ooh this one gets a little sticky when we keep optional locales in mind....

i can imagine something like:
```suggestion
    if (this.shouldUseUtc && this.shouldUseLocale) {
      return dayjs(input, { format: format, locale: locale, utc: this.shouldUseUtc }, locale).utc()
    }
    if (this.shouldUseUtc) {
      return dayjs.utc(input, format);
    }
    if (this.shouldUseLocale) {
      return dayjs(input, { format: format, locale: locale }, locale);
    }
    return dayjs(input, format);
```

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-formats.ts [hunk 3]
```diff
@@ -5,10 +5,10 @@ import { MatDateFormats } from '@angular/material/core';
  */
 export const MAT_DAYJS_DATE_FORMATS: MatDateFormats = {
   parse: {
-    dateInput: 'MM.DD.YYYY',
+    dateInput: 'L',
```
Comment (tabuckner) [inline projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-formats.ts]: Are we sure this works? Just checked it out on a quick test:
https://codepen.io/SwagGrenade/pen/eYmjdqV?editors=1111

image
image

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts [hunk 4]
```diff
@@ -135,16 +143,18 @@ export class DayjsDateAdapter extends DateAdapter<Dayjs> {
...
-    return value ? this.dayJs(value, parseFormat) : null;
+    if (value && typeof value == 'string') {
+      return this.dayJs(value, dayjs().localeData().longDateFormat(parseFormat), this.locale);
+    }
+    return value ? this.dayJs(value).locale(this.locale) : null;
   }
 
   format(date: Dayjs, displayFormat: string): string {
     if (!this.isValid(date)) {
-      throw Error('MomentDateAdapter: Cannot format invalid date.');
```
[the remark this answers is not in the document]
Comment (vanrossumict) [reply projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts]: Haha ah well: better copy paste something that is good than create something by yourself that's totally worse

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts [hunk 5]
```diff
@@ -203,13 +213,12 @@ export class DayjsDateAdapter extends DateAdapter<Dayjs> {
     return this.dayJs(date).isValid();
   }
 
-  // TODO: How to create an invalid Dayjs object?
   invalid(): Dayjs {
-    return;
+    return this.dayJs(null);
```
[the remark this answers is not in the document]
Comment (vanrossumict) [reply projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts]: Haha this was an "educated guess" 😂
I have not yet found a testcase where it calls this method.

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts [hunk 6]
```diff
@@ -55,21 +58,26 @@ export class DayjsDateAdapter extends DateAdapter<Dayjs> {
     if (this.shouldUseUtc) {
       dayjs.extend(utc);
     }
+
+    dayjs.extend(LocalizedFormat);
```
[the remark this answers is not in the document]
Comment (vanrossumict) [reply projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts]: Your absolutely right. I didn't want to go that far just yet.

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts (hunk 1 above)
Comment (vanrossumict) [reply projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts]: The locale is set by default to "en_US" by Angular Material. 
So checking locale for falsiness will not be necessary.

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts (hunk 2 above)
Comment (vanrossumict) [reply projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-adapter.ts]: See comment before: locale will always be set so it shouldn't be a problem.
If I'm not mistaken you can pass the locale parameter without including the localization plugin.
We can try that later.

File: projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-formats.ts (hunk 3 above)
Comment (vanrossumict) [reply projects/material-dayjs-adapter/src/lib/adapter/dayjs-date-formats.ts]: To use "L" (which means localized date format with leading zero) you have to include/extend Day.js with the localization plugin. 
Furthermore, to parse a string it with "L" format, you need to include/extend the customParseFormat plugin.

**Comment by vanrossumict** [comment]
Hi tabuckner,
It was a pleasure
On the one hand I also think adding a shouldUseLocale option to the configuration is a good plan. I was also thinking about that. You can keep the bundle size insanely small.
The solution for this might be really simple: instead of the "L" date fromat, you can use "MM/DD/YYYY" which is the default format of the en_US locale that is set by Angular Material. And of course some way of lazy loading the plugins. Btw: you can override this format by injecting your own MAT_DAYJS_DATE_FORMATS right?
On the other hand, this might overcomplicate things. I have noticed that adding these 3 extra plugins do not increase the bundle size that much. The plugins are also pretty lean. The size starts to really grow when you start adding locales but that's outside of this library.
Let me know what you think.
I might have some time this weekend to do some coding and testing.
Looking forward to your suggestions!

**Comment by tabuckner** [comment]
> The solution for this might be really simple: instead of the "L" date fromat, you can use "MM/DD/YYYY" which is the default format of the en_US locale that is set by Angular Material. And of course some way of lazy loading the plugins. Btw: you can override this format by injecting your own MAT_DAYJS_DATE_FORMATS right?
YES! Great idea. Implementing today.
> On the other hand, this might overcomplicate things. I have noticed that adding these 3 extra plugins do not increase the bundle size that much. The plugins are also pretty lean. The size starts to really grow when you start adding locales but that's outside of this library.
> Let me know what you think.
Okay, I'll trust you on this. Will merge the remaining changes today, and then work on the next step that I think we're on the same page for. 

@vanrossumict Dude, seriously. Thank you for this work--awesome!

**Comment by vanrossumict** [comment]
You're most welcome.
I hope to have some time soon to finish this off.
Btw: I've request this project to be listed on the documentation page of Angular Material, let's hope they do it so this project gets some more attention: https://github.com/angular/components/issues/18337

**Pull request merged**